### ISSUE TITLE
INDArray.mmul numerically unstable

### ISSUE THREAD

**Issue by GrgrLmml** [deeplearning4j/deeplearning4j#6788]
INDArray.mmul applied to a vector and its transposed should result in a symmetric matrix.

   `INDArray rnd = Nd4j.rand(180, 300);`
   `INDArray mmul = rnd.mmul(rnd.transpose());`

It is expected that 
  `mmul.getDouble(i,j) == mmul.getDouble(i,j)`
holds for all `i,j`

This is, however, sometimes not the case. There error is usually in the range of 10^-6 
I tested this with version 0.9.1 through 1.0.0-beta3 on macOS and linux using the CPU backend. 

I prepared a gist
[link to gist](https://gist.github.com/GrgrLmml/888f46afc3206a7d6076f6ee2655f8f0#file-matmultest-java)

**Comment by agibsonccc** [comment]
@GrgrLmml could you tell us which blas you're using? A small self contained project would help.
You can use mkl or cuda. We just call out to external blas routines there.
Have you tried using double as a data type instead?

**Comment by raver119** [comment]
We're not implementing own gemm, there's openblas used in 0.9.1 and MKL in beta3... 

Plus, there's float -> double conversion applied if your dtype isn't double.  By default dtype is float.

**Comment by GrgrLmml** [comment]
Thanks for your prompt reply

tried with `mmul.getFloat(i,j);` same result

This is my setup
<img width="606" alt="untitled" src="https://user-images.githubusercontent.com/5629076/49209038-ab1b6a00-f3b9-11e8-91dc-e8903572c8d3.png">

I also created a self contained project. [project](https://github.com/GrgrLmml/dl4jplayground)

But if I understood you right then this issues has nothing todo with nd4j but with the BLAS library I am using. Is this correct?

**Comment by treo** [comment]
Basically, you are using float precision when calculating mmul, so you get less precision than if you used double.

```
        Nd4j.setDataType(DataBuffer.Type.DOUBLE);
        INDArray rnd = Nd4j.rand(180, 300);
        INDArray mmul = rnd.mmul(rnd.transpose());

        System.out.println(Transforms.abs(mmul.sub(mmul.transpose())).maxNumber());
```
That will output: `0.0`

while this:
```
        INDArray rnd = Nd4j.rand(180, 300);
        INDArray mmul = rnd.mmul(rnd.transpose());

        System.out.println(Transforms.abs(mmul.sub(mmul.transpose())).maxNumber());
```
This will output something in the `4.5E-5` to `6E-5` range.

**Comment by GrgrLmml** [comment]
` Nd4j.setDataType(DataBuffer.Type.DOUBLE);` indeed solves the issue for 1.0.0-beta3 and 0.9.1. 
However in 1.0.0-alpha  even

    Nd4j.setDataType(DataBuffer.Type.FLOAT);
    INDArray rnd = Nd4j.rand(180, 300);
    INDArray mmul = rnd.mmul(rnd.transpose());
    System.out.println(Transforms.abs(mmul.sub(mmul.transpose())).maxNumber());

outputs `0.0` 

So maybe in 1.0.0-alpha there was the data type hardcoded to `Double` and `Nd4j.setDataType(DataBuffer.Type.FLOAT);` 
I first saw the numerical instability when I switched to 1.0.0-beta3 from 1.0.0-alpha. 
`Nd4j.setDataType(DataBuffer.Type.DOUBLE);`  solves it for me.

Thanks for your help
